### PR TITLE
feat: add analytics user profile - Part 2 (WPB-8978)

### PR DESCRIPTION
**Pull request by alexandreferris**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/WPB-8978" title="WPB-8978" target="_blank">TaskWPB-8978</a>  [Android] Countly analytics ID and user properties
  </td></table>
  <br />
 

----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [X] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [X] contains a reference JIRA issue number like `SQPIT-764`
  - [X] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [X] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

### Issues

There was no way of observing when either a new tracking identifier was received or it needed to be generated

### Causes (Optional)

Not implemented.

### Solutions

This is Part 2, which solves Observing if:
- there is a receiving tracking identifier from another client.
- there is already a tracking identifier in use for current client.
- or if there is none, generating a new tracking identifier. (sending part will be done from AR)

### Dependencies (Optional)

- [X] #2868 

#### Test Coverage (Optional)

- [X] I have added automated test to this contribution

### Notes (Optional)

This is Part 2

### CODE REVIEW

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/analytics/ObserveAnalyticsTrackingIdentifierStatusUseCase.kt [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+
+    // TODO(Analytics): Just set identifier without migration
+    data class ExistingIdentifier(
+        override val identifier: String
+    ) : AnalyticsIdentifierResult()
+
+    // TODO(Analytics): Set identifier with migration
+    data class MigrationIdentifier(
+        override val identifier: String
+    ) : AnalyticsIdentifierResult()
```
Comment (alexandreferris) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/analytics/ObserveAnalyticsTrackingIdentifierStatusUseCase.kt:97]: TODOs will be removed as soon as the flow is tested via AR

Comment (alexandreferris) [reply logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/analytics/ObserveAnalyticsTrackingIdentifierStatusUseCase.kt]: Removed and moved this sealed class to its own file inside `data` module.

**Review by vitorhugods — APPROVED** [review]

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/configuration/UserConfigRepository.kt [hunk 2]
```diff
@@ -137,6 +137,11 @@ interface UserConfigRepository {
     fun setShouldFetchE2EITrustAnchors(shouldFetch: Boolean)
     fun getShouldFetchE2EITrustAnchor(): Boolean
     suspend fun setTrackingIdentifier(identifier: String)
+    suspend fun getTrackingIdentifier(): String?
+    suspend fun observeTrackingIdentifier(): Flow<Either<StorageFailure, String>>
+    suspend fun setPreviousTrackingIdentifier(identifier: String)
+    suspend fun getPreviousTrackingIdentifier(): String?
```
[the remark this answers is not in the document]
Comment (alexandreferris) [reply logic/src/commonMain/kotlin/com/wire/kalium/logic/configuration/UserConfigRepository.kt:143]: Yes, we set the `PreviouTrackingIdentifier` when we receive a new `TrackingIdentifier` from remote (meaning a new device of the user was logged in somewhere else and we now need to use this new tracking identifier).

Both identifiers are kept until a proper analytics migration of the new identifier is done, then the previous identifier is deleted.

File: data/src/commonMain/kotlin/com/wire/kalium/logic/data/analytics/AnalyticsIdentifierResult.kt [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see http://www.gnu.org/licenses/.
+ */
+
+package com.wire.kalium.logic.data.analytics
+
+sealed interface AnalyticsIdentifierResult {
```
Comment (yamilmedina) [inline data/src/commonMain/kotlin/com/wire/kalium/logic/data/analytics/AnalyticsIdentifierResult.kt:21]: Suggestion: Could be cool to add some comments here, about what are is purpose of each type 🙌

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/configuration/UserConfigRepository.kt [hunk 4]
```diff
@@ -137,6 +137,11 @@ interface UserConfigRepository {
     fun setShouldFetchE2EITrustAnchors(shouldFetch: Boolean)
     fun getShouldFetchE2EITrustAnchor(): Boolean
     suspend fun setTrackingIdentifier(identifier: String)
+    suspend fun getTrackingIdentifier(): String?
+    suspend fun observeTrackingIdentifier(): Flow<Either<StorageFailure, String>>
```
Comment (yamilmedina) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/configuration/UserConfigRepository.kt:141]:
```suggestion
    suspend fun setCurrentTrackingIdentifier(newIdentifier: String)
    suspend fun getCurrentTrackingIdentifier(): String?
    suspend fun observeCurrentTrackingIdentifier(): Flow<Either<StorageFailure, String>>
```

Question: I like it and I understand it, but, What if... we do something like this ☝️  would it make it clearer 🤔 
(adding the Current word to it) ?

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/analytics/AnalyticsIdentifierManager.kt [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+import com.wire.kalium.logic.data.user.UserId
+import com.wire.kalium.logic.feature.message.MessageSender
+import com.wire.kalium.logic.functional.flatMap
+import com.wire.kalium.logic.functional.foldToEitherWhileRight
+import com.wire.kalium.logic.kaliumLogger
+import kotlinx.datetime.Clock
+
+interface AnalyticsIdentifierManager {
+
+    suspend fun onMigrationComplete()
```
Comment (yamilmedina) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/analytics/AnalyticsIdentifierManager.kt:37]: Suggestion: maybe we can add docs here for this guy in particular, about what the migration is ? :D

**Review by yamilmedina — COMMENTED** [review]
Great work as always 👏  added some suggestions about naming and docs, that can help us in the future, looking sharp 🔪

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/configuration/UserConfigRepository.kt (hunk 4 above)
Comment (alexandreferris) [reply logic/src/commonMain/kotlin/com/wire/kalium/logic/configuration/UserConfigRepository.kt:141]: Oh for sure!

**Review by yamilmedina — APPROVED** [review]

**Pull request merged**